src/kdc/kdc_preauth.c

```diff
@@ -820,20 +811,26 @@ hint_list_next(struct hint_state *arg);
 static void
 finish_get_edata(void *arg, krb5_error_code code, krb5_pa_data *pa)
 {
+    krb5_error_code ret;
     struct hint_state *state = arg;
 
     if (code == 0) {
         if (pa == NULL) {
-            /* Include an empty value of the current type. */
-            pa = calloc(1, sizeof(*pa));
-            pa->magic = KV5M_PA_DATA;
-            pa->pa_type = state->pa_type;
+            ret = alloc_pa_data(state->pa_type, 0, &pa);
+            if (ret)
+                goto error;
         }
-        *state->pa_cur++ = pa;
+        ret = add_pa_data_element(&state->pa_data, pa);
+        if (ret)
+            goto error;
```

Comment:
I think this leaks `pa` (if `code ==0` but `pa == NULL`).

Reply:
Since add_pa_data_element() claims pa, I had it do so on success and failure for simplicity.  (Having it copy so that it never claims pa would be more reasonable for a library API, but within the KDC I figure the looser, more efficient memory management is acceptable as long as it's consistent between success and failure.)

Reply:
I decided that perhaps designing a list accumulator to claim caller memory should carry the cost of a comment at each call site.  Ordinarily I don't like documenting function behavior at every call site, but unusual memory management may be an exception.  So I've added comments like ``/* add_pa_data_element() claims pa on success or failure. */`` to each call.